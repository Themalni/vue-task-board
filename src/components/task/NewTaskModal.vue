<template>
  <Modal title="Stwórz nowe zadanie" :status="task.status">
    <template v-slot:body>
      <div class="container pb-3">
        <h4>Opis zadania:</h4>
        <textarea
          class="mt-2"
          :class="[{ 'has-error': errors.first('description') }]"
          name="description"
          type="textarea"
          label="Description"
          v-model="task.description"
          v-validate="'required'"
          data-vv-as="Opis zadania"
          placeholder="Dodaj opis zadania..."
          rows="6"
          @focus="isEditing = true"
        ></textarea>
        <small
          v-if="errors.first('description')"
          class="text-xs error-text edit-error"
        >{{ errors.first('description') }}</small>
      </div>
        <h4 class="w-100">Rodzaj zadania:</h4>
        <Select
          :name="'taskTypesSelect'"
          id="select"
          label
          v-model="task.type"
          rules="required"
          friendly-name="Rodzaj zadania"
          :options="taskTypes"
          default-option="Wybierz rodzaj zadania"
        ></Select>
    </template>
    <template v-slot:footer>
      <Button color="button-primary" :disabled="!isEditing" @addAction="saveChanges()">Zachowaj</Button>
      <Button class="ml-2" color="button-default" @addAction="closeModal()">Anuluj</Button>
    </template>
  </Modal>
</template>

<script>
import Modal from '@/components/Modal'
import Button from '@/elements/Button'
import Select from '@/elements/Select'
import { mapActions } from 'vuex'

export default {
  name: 'NewTaskModal',
  inject: ['$validator'],
  components: {
    Modal,
    Button,
    Select
  },
  data: () => ({
    task: {
      id: null,
      categoryId: 'c1',
      description: null,
      type: null,
      status: 'Oczekujące'
    },
    taskTypes: {
      feature: 'feature',
      bugfix: 'bugfix',
      default: 'default'
    },
    isEditing: false,
    isValid: true
  }),
  methods: {
    ...mapActions(['changeNewTaskModalState', 'saveTask']),
    saveChanges () {
      this.$validator.validate()
        .then(valid => {
          if (valid) {
            this.$store.dispatch('saveTask', this.task)
            this.$store.dispatch('changeNewTaskModalState', false)
            this.isEditing = false
          }
        })
        .catch(() => {
          console.log(this.errors.all())
        })
      console.log(this.errors.all())
    },
    closeModal () {
      this.$store.dispatch('changeNewTaskModalState', false)
    }
  }
}
</script>

<style lang="scss">
.container {
  width: 100%;
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  position: relative;
}
.edit-error {
  position: absolute;
  bottom: 0.5em;
}
textarea {
  padding: 1em;
  border: 1px dashed $grey;
  outline: none;
  resize: none;
  font-family: "Noto Sans", sans-serif;
  font-size: 0.8em;
  color: $graphite;
  border-radius: 2px;
  width: 100%;

  &.has-error {
    border-color: $crimson;
  }
}
</style>
