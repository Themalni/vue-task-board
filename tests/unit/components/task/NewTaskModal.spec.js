import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import VeeValidate from 'vee-validate'
import flushPromises from 'flush-promises'
import mockState from '../../utils/mockState'
import NewTaskModal from '@/components/task/NewTaskModal'

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.use(VeeValidate)

describe('NewTaskModal', () => {
  let wrapper
  let store
  let actions
  let state
  let getters

  let mockError = 'This field is required'

  beforeEach(() => {
    state = mockState
    actions = {
      changeNewTaskModalState: jest.fn(),
      saveTask: jest.fn()
    }
    store = new Vuex.Store({
      state,
      getters,
      actions
    })

    wrapper = mount(NewTaskModal, {
      setData: ({
        task: {
          id: null,
          categoryId: 'c1',
          description: null,
          type: null,
          status: 'OczekujÄ…ce'
        },
        taskTypes: {
          feature: 'feature',
          bugfix: 'bugfix',
          default: 'default'
        },
        isEditing: false,
        isValid: true
      }),
      store,
      localVue,
      sync: false
    })
  })

  afterEach(() => {
    wrapper.destroy()
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('calls `closeModal` when button is clicked', () => {
    wrapper.vm.closeModal()
    expect(actions.changeNewTaskModalState.mock.calls).toHaveLength(1)
  })
  it('calls `saveChanges` and validated inputs when button is clicked', async () => {
    wrapper.vm.$validator.validate = jest.fn(() => Promise.resolve(true))

    wrapper.vm.saveChanges()

    await flushPromises()

    expect(wrapper.vm.$validator.validate).toHaveBeenCalled()

    expect(actions.saveTask.mock.calls).toHaveLength(1)
    expect(actions.changeNewTaskModalState.mock.calls).toHaveLength(1)
    expect(wrapper.vm.isEditing).toBe(false)
  })
  it('catches an error during validation', async () => {
    wrapper.vm.editing = true
    wrapper.vm.$validator.validate = jest.fn(() => Promise.reject(mockError))

    wrapper.vm.saveChanges()

    await flushPromises()

    expect(wrapper.vm.$validator.validate).toHaveBeenCalled()
    expect(mockError).toBe('This field is required')
  })
})
