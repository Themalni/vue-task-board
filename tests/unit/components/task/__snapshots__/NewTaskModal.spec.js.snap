// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewTaskModal renders correctly 1`] = `
<div class="overlay" name="fade">
  <div class="modal">
    <div class="modal-header">
      <h3 class="modal-header-text text-white">Stwórz nowe zadanie</h3>
      <h3 class="modal-header-text text-white ml-1">(Oczekujące)</h3>
    </div>
    <div class="modal-body">
      <div class="container pb-3">
        <h4>Opis zadania:</h4> <textarea name="description" type="textarea" label="Description" data-vv-as="Opis zadania" placeholder="Dodaj opis zadania..." rows="6" class="mt-2" aria-required="true" aria-invalid="false"></textarea>
        <!---->
      </div>
      <h4 class="w-100">Rodzaj zadania:</h4>
      <div class="select-container pb-3"><label>

          <select name="taskTypesSelect" data-vv-as="Rodzaj zadania" data-vv-validate-on="blur" aria-required="true" aria-invalid="false">
            <option value="">Wybierz rodzaj zadania</option>
            <option value="feature">feature</option>
            <option value="bugfix">bugfix</option>
            <option value="default">default</option>
          </select></label>
        <!---->
      </div>
    </div>
    <div class="modal-footer">
      <div><button disabled="disabled" class="button button-primary">Zachowaj</button></div>
      <div class="ml-2"><button class="button button-default">Anuluj</button></div>
    </div>
  </div>
</div>
`;
